src/drivers/imx/edma.c

```diff
@@ -4,109 +4,858 @@
+	// Reenable HW requests
+//	edma_chan_update_bits(channel, EDMA_CH_CSR,
+//			      EDMA_CH_CSR_ERQ_EARQ, EDMA_CH_CSR_ERQ_EARQ);
 	return 0;
 }
 
 static int edma_pause(struct dma_chan_data *channel)
 {
+	// TODO Why is this function here, given that it is never called?
+	tracev_edma("EDMA: pause(%d)", channel->index);
+	// Validate state
+	switch (channel->status) {
+	case COMP_STATE_ACTIVE:
+		break;
+	default:
+		return -EINVAL;
+	}
+	channel->status = COMP_STATE_PAUSED;
+	// Disable HW requests
+	edma_chan_update_bits(channel, EDMA_CH_CSR,
+			      EDMA_CH_CSR_ERQ_EARQ, 0);
 	return 0;
 }
 
 static int edma_stop(struct dma_chan_data *channel)
 {
+	tracev_edma("EDMA: stop(%d)", channel->index);
+	// Validate state
+	// TODO: Should we?
+	switch (channel->status) {
+	case COMP_STATE_READY:
+	case COMP_STATE_PREPARE:
+		return 1; /* Already stopped, don't propagate request */
+	case COMP_STATE_PAUSED:
+	case COMP_STATE_ACTIVE:
+		break;
+	default:
+		return -EINVAL;
+	}
+	channel->status = COMP_STATE_READY;
+	/* Disable channel */
+	edma_chan_write(channel, EDMA_CH_CSR, 0);
+	edma_chan_write(channel, EDMA_TCD_CSR, 0);
+	/* The remaining TCD values will still be valid but I don't care
+	 * about them anyway
+	 */
+
+	return 0;
+}
+
+static int edma_copy(struct dma_chan_data *channel, int bytes, uint32_t flags)
+{
+	struct dma_cb_data next = { .elem.size = bytes };
+
+	tracev_edma("edma_copy() is a nop to me");
+	/* We need to call the copy() callback */
+	if (channel->cb && channel->cb_type & DMA_CB_TYPE_COPY)
+		channel->cb(channel->cb_data, DMA_CB_TYPE_COPY, &next);
 	return 0;
 }
 
 static int edma_status(struct dma_chan_data *channel,
 		       struct dma_chan_status *status, uint8_t direction)
 {
+	status->state = channel->status;
+	status->flags = 0;
+	/* Note: these might be slightly inaccurate as they are only
+	 * updated at the end of each minor (block) transfer
+	 */
+	status->r_pos = edma_chan_read(channel, EDMA_TCD_SADDR);
+	status->w_pos = edma_chan_read(channel, EDMA_TCD_DADDR);
+	status->timestamp = timer_get_system(platform_timer);
 	return 0;
 }
 
+static inline int signum(int x)
+{
+	return x < 0 ? -1 : x == 0 ? 0 : 1;
+}
+
+static int validate_nonsg_config(struct dma_sg_elem_array *sgelems,
+				 uint32_t SOFF, uint32_t DOFF)
+{
+	uint32_t sbase, dbase, size;
+
+	if (!sgelems)
+		return -EINVAL;
+	if (sgelems->count != 2)
+		return -EINVAL; /* Only ping-pong configs supported */
+
+	sbase = sgelems->elems[0].src;
+	dbase = sgelems->elems[0].dest;
+	size = sgelems->elems[0].size;
+
+	if (sbase + size * signum(SOFF) != sgelems->elems[1].src)
+		return -EINVAL; /**< Not contiguous */
+	if (dbase + size * signum(DOFF) != sgelems->elems[1].dest)
+		return -EINVAL; /**< Not contiguous */
+	if (size != sgelems->elems[1].size)
+		return -EINVAL; /**< Mismatched sizes */
+
+	return 0; /* Ok, we good */
+}
+
+/* Some set_config helper functions */
+static int setup_tcd(struct dma_chan_data *channel, uint16_t SOFF,
```

Comment:
add a proper comment to this function. Avoid uppercase names.

Reply:
All addressed.